src/Aspire.Hosting.Redis/RedisBuilderExtensions.cs

```diff
@@ -24,13 +24,72 @@ public static class RedisBuilderExtensions
-                      .WithImageRegistry(RedisContainerImageTags.Registry);
+        return builder.AddRedis(name, port, null);
+    }
+
+    /// <summary>
+    /// Adds a Redis container to the application model.
+    /// </summary>
+    /// <remarks>
+    /// The default image is "redis" and the tag is "7.2.4".
+    /// </remarks>
+    /// <param name="builder">The <see cref="IDistributedApplicationBuilder"/>.</param>
+    /// <param name="name">The name of the resource. This name will be used as the connection string name when referenced in a dependency.</param>
+    /// <param name="port">The host port to bind the underlying container to.</param>
+    /// <param name="password">The parameter used to provide the password for the Redis resource. If <see langword="null"/> resource will be not password protected.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<RedisResource> AddRedis(
+        this IDistributedApplicationBuilder builder,
+        string name,
+        int? port = null,
+        IResourceBuilder<ParameterResource>? password = null)
+    {
+
+        if (password?.Resource is not null)
+        {
+
+            var redis = new RedisResource(name, password.Resource);
+            return builder.AddResource(redis)
+                          .WithEndpoint(port: port, targetPort: 6379, name: RedisResource.PrimaryEndpointName)
+                          .WithImage(RedisContainerImageTags.Image, RedisContainerImageTags.Tag)
+                          .WithImageRegistry(RedisContainerImageTags.Registry)
+                          .WithPassword(password.Resource);
+        }
+        else
+        {
+            if (builder.ExecutionContext.IsPublishMode)
+            {
+                var passwordParameter = password?.Resource ?? ParameterResourceBuilderExtensions.CreateDefaultPasswordParameter(builder, $"{name}-password");
+                var redis = new RedisResource(name, passwordParameter);
+
+                return builder.AddResource(redis)
+                          .WithEndpoint(port: port, targetPort: 6379, name: RedisResource.PrimaryEndpointName)
+                          .WithImage(RedisContainerImageTags.Image, RedisContainerImageTags.Tag)
+                          .WithImageRegistry(RedisContainerImageTags.Registry)
+                          .WithPassword(passwordParameter);
+            }
+            else if (builder.ExecutionContext.IsRunMode)
+            {
+                var redis = new RedisResource(name);
+                return builder.AddResource(redis)
+                              .WithEndpoint(port: port, targetPort: 6379, name: RedisResource.PrimaryEndpointName)
+                              .WithImage(RedisContainerImageTags.Image, RedisContainerImageTags.Tag)
+                              .WithImageRegistry(RedisContainerImageTags.Registry);
+            }
+            throw new InvalidOperationException();
```

Comment:
Can we add a message to this exception to explain what went wrong?

Reply:
It is a shame that we have to do these parameters in this order (its different to `AddPostgres`).